### ISSUE TITLE
Can't pull image from cluster unauthorized

### ISSUE THREAD

**Issue by dominic-p** [ContainerSolutions/trow#251]
**Describe the bug**
I'm having trouble pulling images from within the cluster. I can push and pull images from my workstation no problem, but anytime I try to pull an image from a node in the cluster I get:

> FATA[0000] pulling image: rpc error: code = Unknown desc = unable to retrieve auth token: invalid username/password: unauthorized: authentication required

**To Reproduce**
1. Install trow with minimal components (see below) and push an image
2. Attempt to pull that image either via creating a deployment or with crictl

**Expected behavior**
The image should pull when given the proper credentials.

**Output/Logs**
The trow container logs look like this:
```
[2021-05-31T18:08:47Z WARN  rocket::rocket] Responding with 401 Unauthorized catcher.
[2021-05-31T18:09:41Z WARN  rocket::rocket] Responding with 401 Unauthorized catcher.
[2021-05-31T18:09:41Z WARN  rocket::rocket] Responding with 401 Unauthorized catcher.
```

**Trow Info**
 - Install method (quick-install, Helm, Kustomize): I want to use this exclusively for in-cluster testing, so I'm not bothering with Ingress or SSL. I just grabbed the yaml for the StatefulSet and the Service [from here](https://github.com/ContainerSolutions/trow/tree/v0.3.2/install/base) and tweaked it a bit. I used `kubectl port-forward` to connect from my workstation and after logging in I was able to push and pull images as described in the [quick install docs](https://github.com/ContainerSolutions/trow/blob/v0.3.2/QUICK-INSTALL.md#test-it-out).

Notably, my args for the trow container look like this:

```
        args:
          - "--no-tls"
          - "-n"
          - "trow-svc.trow.svc.cluster.local:8000"
          - "-u"
          - "test"
          - "--password-file"
          - "/etc/trow/pass"
```

I added the registry on each node to /etc/containers/registries.conf like this:

```
[[registry]]
prefix = "trow-svc.trow.svc.cluster.local:8000"
insecure = true
blocked = false
location = "<cluster ip of service>:8000"
```

 - Version/Image Name: `containersol/trow:0.3`

**Kubernetes**
 - Kubernetes distro/host (EKS, GKE, Kind, Minikube etc): vanilla k8s managed by kubeadm on raspberry pi 4 cluster arm64 Ubuntu with CRI-O
 - Kubernetes version (e.g. 1.19): 1.19.11

**Additional context**
I've tried testing with `imagePullSecrets:` and with `crictl pull --creds test:password trow-svc.trow.cluster.local:8000/test/nginx` and I get the same unauthorized result either way.

My gut is telling me this may be a compatibility issue with CRI-O since podman can pull the image no problem, but I can't understand what the issue could be. Is there any way I can get more verbose logging?

**Comment by dominic-p** [comment]
I cranked up CRI-O's log level and got this in the logs when attempting to pull the image via `crictl`:

```
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.969581483-07:00" level=debug msg="Request: &PullImageRequest{Image:&ImageSpec{Image:trow-svc.trow.svc.cluster.local:8000/test/nginx,Annotations:map[string]string{},},Auth:&AuthConfig{Username:,Password:,Auth:dGVzdDpuOVdqUlh4RURiTk5jSTB4c0RHOGdDQ2VZZnc9,ServerAddress:,IdentityToken:,RegistryToken:,},SandboxConfig:nil,}" file="go-grpc-middleware/chain.go:25" id=93e7a7a8-e4c7-4b22-8a31-d88c239b9597 name=/runtime.v1alpha2.ImageService/PullImage
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.972104104-07:00" level=info msg="Pulling image: trow-svc.trow.svc.cluster.local:8000/test/nginx" file="server/image_pull.go:32" id=93e7a7a8-e4c7-4b22-8a31-d88c239b9597 name=/runtime.v1alpha2.ImageService/PullImage
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.974239155-07:00" level=debug msg="Skipping non-existing decryption_keys_path: /etc/crio/keys/" file="server/utils.go:238"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.979997576-07:00" level=debug msg="Trying to access \"<cluster ip of service>:8000/test/nginx:latest\"" file="docker/docker_image_src.go:68"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.981637355-07:00" level=debug msg="Credentials not found" file="config/config.go:179"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.982684770-07:00" level=debug msg="Using registries.d directory /etc/containers/registries.d for sigstore configuration" file="docker/lookaside.go:51"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.984180661-07:00" level=debug msg=" Using \"default-docker\" configuration" file="docker/lookaside.go:169"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.988390172-07:00" level=debug msg=" No signature storage configuration found for <cluster ip of service>:8000/test/nginx:latest" file="docker/lookaside.go:174"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.990124172-07:00" level=debug msg="Looking for TLS certificates and private keys in /etc/docker/certs.d/<cluster ip of service>:8000" file="tlsclientconfig/tlsclientconfig.go:21"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.991927746-07:00" level=debug msg="GET https://<cluster ip of service>:8000/v2/" file="docker/docker_client.go:516"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.996243255-07:00" level=debug msg="Ping https://<cluster ip of service>:8000/v2/ err Get \"https://<cluster ip of service>:8000/v2/\": EOF (&url.Error{Op:\"Get\", URL:\"https://<cluster ip of service>:8000/v2/\", Err:(*errors.errorString)(0x400007e060)})" file="docker/docker_client.go:708"
May 31 17:11:40 node-1 crio[2099910]: time="2021-05-31 17:11:40.996601714-07:00" level=debug msg="GET http://<cluster ip of service>:8000/v2/" file="docker/docker_client.go:516"
May 31 17:11:41 node-1 crio[2099910]: time="2021-05-31 17:11:41.000246565-07:00" level=debug msg="Ping http://<cluster ip of service>:8000/v2/ status 401" file="docker/docker_client.go:712"
May 31 17:11:41 node-1 crio[2099910]: time="2021-05-31 17:11:41.001186333-07:00" level=debug msg="GET http://<cluster ip of service>:8000/login?scope=repository%3Atest%2Fnginx%3Apull&service=trow_registry" file="docker/docker_client.go:675"
May 31 17:11:41 node-1 crio[2099910]: time="2021-05-31 17:11:41.006577090-07:00" level=debug msg="Accessing \"<cluster ip of service>:8000/test/nginx:latest\" failed: unable to retrieve auth token:invalid username/password: unauthorized: authentication required" file="docker/docker_image_src.go:73"
May 31 17:11:41 node-1 crio[2099910]: time="2021-05-31 17:11:41.006851698-07:00" level=debug msg="Error preparing image trow-svc.trow.svc.cluster.local:8000/test/nginx: unable to retrieve auth token: invalid username/password: unauthorized: authentication required" file="server/image_pull.go:134" id=93e7a7a8-e4c7-4b22-8a31-d88c239b9597 name=/runtime.v1alpha2.ImageService/PullImage
May 31 17:11:41 node-1 crio[2099910]: time="2021-05-31 17:11:41.007088640-07:00" level=debug msg="Response error: unable to retrieve auth token: invalid username/password: unauthorized: authentication required" file="go-grpc-middleware/chain.go:25" id=93e7a7a8-e4c7-4b22-8a31-d88c239b9597 name=/runtime.v1alpha2.ImageService/PullImage
```

I'm not entirely sure what to make of this, but we can clearly see trow responding with a 401 error and then a call to /login that fails for some reason.

**Comment by amouat** [comment]
Thanks. `imagePullSecrets` should work. I'm afraid I'm flat out this week, but I'll look into this next week.

**Comment by amouat** [comment]
I'm taking a look today by using minikube with CRI-O. But in the meantime, if you want more logs, try setting the RUST_LOG env variable. I *think* it should work with the executable in the container, but it's possible you'll need a debug build. It should look something like:

```
RUST_LOG=trace ./trow --no-tls ...
```
Trace is probably very noisy, info might be better.

**Comment by dominic-p** [comment]
Thanks for digging into this. I tried adding RUST_LOG=trace to the container spec for my trow StatefulSet, but unfortunately, I'm not seeing any additional logs. I'm just getting:

```
[2021-06-10T04:18:04Z WARN  rocket::rocket] Responding with 401 Unauthorized catcher.
```

Let me know if there's any other debugging I can help with.

**Comment by amouat** [comment]
I've been trying to recreate this and failing - it works for me. I'm running minikube on my host with:

```
$ minikube start --container-runtime=crio
```
Which is using:

```
Kubernetes v1.20.2 on CRI-O 1.20.2
```
It works with both Kubernetes `imagePullSecrets` and using crictl after using `minikube ssh`:

```
$ sudo crictl --debug pull --creds test:test 10.244.0.3:8000/alpine
DEBU[0000] get image connection                         
DEBU[0000] connect using endpoint 'unix:///var/run/crio/crio.sock' with '2s' timeout 
DEBU[0000] connected successfully using endpoint: unix:///var/run/crio/crio.sock 
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:10.244.0.3:8000/alpine,Annotations:map[string]string{},},Auth:&AuthConfig{Username:test,Password:test,Auth:,ServerAddress:,IdentityToken:,RegistryToken:,},SandboxConfig:nil,} 
DEBU[0000] PullImageResponse: &PullImageResponse{ImageRef:10.244.0.3:8000/alpine@sha256:12e1bb1bfc76415c157e352f69c3476a9da3187a05b8b2307f42aab8adf6f994,} 
Image is up to date for 10.244.0.3:8000/alpine@sha256:12e1bb1bfc76415c157e352f69c3476a9da3187a05b8b2307f42aab8adf6f994
``` 

Trow is running with `--no-tls`. I just hardcoded the service address everywhere, which works in this case because the minikube "host" is in the same network namespace as k8s (not usually the case). 

What's the exact version of Trow your using? (You should see it in the logs).

Let me check and see how to get you a debug build. I'm a bit lost at the minute, but there's still quite a few differences between our setups.

**Comment by amouat** [comment]
To get debugging, you can do the following:

```
     containers:
      - name: trow-pod
        image: containersol/trow:0.3
        command: 
          - "env"
          - "RUST_LOG=trace" 
          - "/trow"
          - "--no-tls" 
          - "-n" 
          - "trow-svc.trow.svc.cluster.local" 
          - "-u"
          - "test"
          - "--password"
          - "test"
```

Note that you need to use `command` not `args` and `env`.

**Comment by amouat** [comment]
One thing I'd ask you to test is using `--password` instead of `--password-file` and then run both Docker and crictl against the registry. 

If nothing is coming up in the logs, we can take things further by running ngrep or similar to grab the traffic and compare the working and breaking requests. I'd really like to be able to recreate the issue though. I'm not sure what to try next on my end.

**Comment by dominic-p** [comment]
Thanks for sticking with me. We have some progress. I tried using `--password` instead of `password-file` and I tried using the service IP instead of the cluster DNS name and I was able to successfully authenticate with `crictl` just like you did.

Then, when I change the command to use the cluster DNS name I get the unauthorized error again. I tried with both `imagePullSecrets:` and `crictl` with the same results.

I was able to get debugging logs using the command you gave me:

```
[2021-06-11T06:08:00Z DEBUG hyper::server] Incoming stream
[2021-06-11T06:08:00Z TRACE hyper::buffer] get_buf []
[2021-06-11T06:08:00Z TRACE hyper::buffer] read_into_buf buf[0..4096]
[2021-06-11T06:08:00Z TRACE hyper::buffer] get_buf [u8; 4096][0..194]
[2021-06-11T06:08:00Z TRACE hyper::http::h1] try_parse([71, 69, 84, 32, 47, 118, 50, 47, 32, 72, 84, 84, 80, 47, 49, 46, 49, 13, 10, 72, 111, 115, 116, 58, 32, 49, 48, 46, 57, 57, 46, 50, 52, 54, 46, 50, 48, 48, 58, 56, 48, 48, 48, 13, 10, 85, 115, 101, 114, 45, 65, 103, 101, 110, 116, 58, 32, 99, 114, 105, 45, 111, 47, 49, 46, 49, 57, 46, 49, 32, 103, 111, 47, 103, 111, 49, 46, 49, 53, 46, 50, 32, 111, 115, 47, 108, 105, 110, 117, 120, 32, 97, 114, 99, 104, 47, 97, 114, 109, 54, 52, 13, 10, 68, 111, 99, 107, 101, 114, 45, 68, 105, 115, 116, 114, 105, 98, 117, 116, 105, 111, 110, 45, 65, 112, 105, 45, 86, 101, 114, 115, 105, 111, 110, 58, 32, 114, 101, 103, 105, 115, 116, 114, 121, 47, 50, 46, 48, 13, 10, 65, 99, 99, 101, 112, 116, 45, 69, 110, 99, 111, 100, 105, 110, 103, 58, 32, 103, 122, 105, 112, 13, 10, 67, 111, 110, 110, 101, 99, 116, 105, 111, 110, 58, 32, 99, 108, 111, 115, 101, 13, 10, 13, 10])
[2021-06-11T06:08:00Z TRACE hyper::http::h1] Request.try_parse([Header; 100], [u8; 194])
[2021-06-11T06:08:00Z TRACE hyper::http::h1] Request.try_parse Complete(194)
[2021-06-11T06:08:00Z TRACE hyper::header] raw header: "Host"=[49, 48, 46, 57, 57, 46, 50, 52, 54, 46, 50, 48, 48, 58, 56, 48, 48, 48]
[2021-06-11T06:08:00Z TRACE hyper::header] raw header: "User-Agent"=[99, 114, 105, 45, 111, 47, 49, 46, 49, 57, 46, 49, 32, 103, 111, 47, 103, 111, 49, 46, 49, 53, 46, 50, 32, 111, 115, 47, 108, 105, 110, 117, 120, 32, 97, 114, 99, 104, 47, 97, 114, 109, 54, 52]
[2021-06-11T06:08:00Z TRACE hyper::header] raw header: "Docker-Distribution-Api-Version"=[114, 101, 103, 105, 115, 116, 114, 121, 47, 50, 46, 48]
[2021-06-11T06:08:00Z TRACE hyper::header] raw header: "Accept-Encoding"=[103, 122, 105, 112]
[2021-06-11T06:08:00Z TRACE hyper::header] raw header: "Connection"=[99, 108, 111, 115, 101]
[2021-06-11T06:08:00Z DEBUG hyper::server::request] Request Line: Get AbsolutePath("/v2/") Http11
[2021-06-11T06:08:00Z DEBUG hyper::server::request] Headers { Host: 10.99.246.200:8000
    , User-Agent: cri-o/1.19.1 go/go1.15.2 os/linux arch/arm64
    , Docker-Distribution-Api-Version: registry/2.0
    , Accept-Encoding: gzip
    , Connection: close
    , }
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.set( "Connection", Connection([Close]) )
[2021-06-11T06:08:00Z TRACE rocket::data::data] Data::new(EmptyReader)
[2021-06-11T06:08:00Z TRACE rocket::data::data] Filled peek buf with 0 bytes.
[2021-06-11T06:08:00Z TRACE rocket::data::data] Peek bytes: 0/512 bytes.
[2021-06-11T06:08:00Z INFO  rocket::rocket] GET /v2/:
[2021-06-11T06:08:00Z TRACE rocket::router] Routing the request: GET /v2/
[2021-06-11T06:08:00Z TRACE rocket::router] All matches: [Route { name: Some("get_v2root"), method: Get, base: Origin { source: None, path: "/", query: None, segment_count: [uninitialized storage] }, uri: Origin { source: None, path: "/v2", query: None, segment_count: [uninitialized storage] }, rank: -4, format: None, metadata: Metadata { path_segments: [RouteSegment { string: "v2", kind: Static, name: "v2", index: Some(0), _part: PhantomData }], query_segments: None, fully_dynamic_query: true } }]
[2021-06-11T06:08:00Z INFO  rocket::rocket] Matched: GET /v2 (get_v2root)
[2021-06-11T06:08:00Z INFO  rocket::rocket] Outcome: Failure
[2021-06-11T06:08:00Z WARN  rocket::rocket] Responding with 401 Unauthorized catcher.
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.append_raw( "www-authenticate", [66, 101, 97, 114, 101, 114, 32, 114, 101, 97, 108, 109, 61, 34, 104, 116, 116, 112, 58, 47, 47, 49, 48, 46, 57, 57, 46, 50, 52, 54, 46, 50, 48, 48, 58, 56, 48, 48, 48, 47, 108, 111, 103, 105, 110, 34, 44, 115, 101, 114, 118, 105, 99, 101, 61, 34, 116, 114, 111, 119, 95, 114, 101, 103, 105, 115, 116, 114, 121, 34, 44, 115, 99, 111, 112, 101, 61, 34, 112, 117, 115, 104, 47, 112, 117, 108, 108, 34] )
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.append_raw( "Content-Type", [97, 112, 112, 108, 105, 99, 97, 116, 105, 111, 110, 47, 106, 115, 111, 110] )
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.append_raw( "Server", [82, 111, 99, 107, 101, 116] )
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.append_raw( "Docker-Distribution-API-Version", [114, 101, 103, 105, 115, 116, 114, 121, 47, 50, 46, 48] )
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.set( "Content-Length", ContentLength(0) )
[2021-06-11T06:08:00Z DEBUG hyper::server::response] writing head: Http11 Unauthorized
[2021-06-11T06:08:00Z TRACE hyper::header] Headers.set( "Date", Date(HttpDate(Tm { tm_sec: 0, tm_min: 8, tm_hour: 6, tm_mday: 11, tm_mon: 5, tm_year: 121, tm_wday: 5, tm_yday: 161, tm_isdst: 0, tm_utcoff: 0, tm_nsec: 268941424 })) )
[2021-06-11T06:08:00Z DEBUG hyper::server::response] headers [
    Headers { Connection: close
    , www-authenticate: Bearer realm="http://10.99.246.200:8000/login",service="trow_registry",scope="push/pull"
    , Content-Type: application/json
    , Server: Rocket
    , Docker-Distribution-API-Version: registry/2.0
    , Content-Length: 0
    , Date: Fri, 11 Jun 2021 06:08:00 GMT
    , }]
[2021-06-11T06:08:00Z TRACE hyper::server::response] ending
[2021-06-11T06:08:00Z INFO  rocket::rocket] Response succeeded.
[2021-06-11T06:08:00Z DEBUG hyper::server] keep_alive = false for 10.244.1.1:44690
[2021-06-11T06:08:00Z DEBUG hyper::server] keep_alive loop ending for 10.244.1.1:44690
```

I'm not really sure how to read those buffer arrays, so I wasn't able to get much useful information, but maybe you can.

So, to recap:

Works
```
sudo crictl --debug pull --creds test:test 10.99.246.200:8000/test/nginx
```

Doesn't work
```
sudo crictl --debug pull --creds test:test trow-svc.trow.svc.cluster.local:8000/test/nginx
```

Any ideas on how to get the domain name version working?

**Comment by dominic-p** [comment]
I just saw that you asked for the version. Saw I missed that before. From the startup logs:

```
Starting Trow 0.3.2 on 0.0.0.0:8000
```

**Comment by amouat** [comment]
The byte arrays aren't very interesting; it's exactly the same content that you see in the Header log line.

Nothing seems to be going wrong in the logs you shared. Trow is just saying, "hey log in here first to get a token" and then the logs end. 

Could you send me separate logs from doing a request with crictl using the IP (should be working) and DNS name (broken) and the exact crictl command you used?

Maybe it's a routing problem somewhere (which could be in Trow).

**Comment by dominic-p** [comment]
Ok, I made a gist with two files. One shows the logs for a successful pull (using the cluster IP) and the other shows the logs for a failed pull (using the cluster DNS name). I included the full command and output as well, so hopefully there's enough information there to go on.

https://gist.github.com/dominic-p/95fd4029204c9d21cb1342d86965eb99

One thing you mentioned kind of confused me:

> Normally hosts (which pull the images) aren't in the k8s network themselves, so they don't understand k8s addresses

I thought the whole point of Trow was to have a container registry running inside the cluster where the images are used. That's how I read [the comparison doc](https://github.com/ContainerSolutions/trow/blob/main/docs/COMPARISON.md). Am I misunderstanding the use case? Or, does "host" in this context mean the kubernetes node itself? I see what you are saying there, but in the case, I would expect a tcp dial error or a domain name couldn't be resolved type error, right?

I tried adding the entry to `/etc/hosts` that you suggested, and unfortunately it had no effect. As far as I know, CRI-O doesn't do any special routing. The registry config is just there, so I can specify `insecure = true`. Otherwise it would only try to connect via HTTPs and fail. We know that `crictl` is able to resolve the address correctly because we are seeing the connection attempt in the trow logs.

Any ideas from the logs?

**Comment by amouat** [comment]
Thanks!

Yes, I can see what's happening reasonably clearly in the logs. If you look at compare lines 83 and 86, you can see that the success logs pass the username and password, but the failure ones do not (it is worth noting that as we're not using TLS the password is passed as base64 encoded plaintext).

The question is why crictl isn't sending the credentials. My only guess is that CRI-O doesn't like that Trow asks it to authenticate to a different address than one it started with (domain name vs IP). I'm not sure if this is 1) a problem in Trow 2) a problem in CRI-O or 3) a config issue.

Regarding what I said about hosts and networking, yes I meant the host/node that k8s is installed on. Normally if you ssh into that node (as opposed to a pod), you won't be able to access internal k8s addresses like the one you're using. It can be frustrating, as containerd/crictl pulls normally also can't resolve internal k8s addresses. This does depend on exactly how you've installed/configured k8s though.

In your case, I wonder if the only reason it's working is because in the registry config you specify: 

```
[[registry]]
prefix = "trow-svc.trow.svc.cluster.local:8000"
insecure = true
blocked = false
location = "<cluster ip of service>:8000"
```

If you don't specify the IP address in location, or change it to the domain name, does it break? What about if you just try to do `curl trow-svc.trow.svc.cluster.local:8000/v2`?

**Comment by amouat** [comment]
Yeah, I just saw this on the CRI-O issue list: https://github.com/cri-o/cri-o/issues/4941

I suspect it's the same issue, although I know you're not using mirrors.

**Comment by dominic-p** [comment]
I think that's it. For some reason, in my mental model since the Pod that runs the container is in the cluster it should be able to pull an image from the cluster.local address. But, it's not the Pod that pulls the image, it's CRI-O which is a service that runs on the host node (not in the cluster), so, of course it can't resolve cluster.local addresses.

I used the [technique here](https://gist.github.com/brasey/fa2277a6d7242cdf4e4b7c720d42b567) to enable the host node's systemd-resolved service to resolve cluster.local addresses. I then modified the entry in /etc/containers/registries.conf to look like this:

```
[[registry]]
insecure = true
blocked = false
location = "trow-svc.trow.svc.cluster.local:8000"
```

And, bingo, I can pull images with:

```
sudo crictl --debug pull --creds test:test trow-svc.trow.svc.cluster.local:8000/test/nginx
```

I also tested `imagePullSecrets:` which works perfectly as well. So, I think we can conclude that the original issue is as you suspected: CRI-O isn't sending credentials because Trow is asking it to authenticate to a different domain than it expects. I have no idea if that is a bug in Trow or CRI-O, but I suspect it's just CRI-O being overly cautious which isn't a bad thing.

In the end, I don't think anything should really change from Trow's perspective. Using cluster.local addresses like this kind of backwards now that I understand things a bit better. I'll need to setup a proper Ingress and/or LoadBalancer for Trow even though I only plan on using it inside the cluster. Sorry to have wasted your time with this one.

**Comment by amouat** [comment]
You didn't waste my time! At worst I learnt something.

I'm considering open an issue/PR on containers/image to log that the user/pass is being stripped. If it was in the logs I think we would have found the problem much earlier. 

I don't like how image pulling is outside of the k8s network, it makes a lot of use cases (like this) unnecessarily difficult.

I'm really not sure what to say about where the issue lies. Trow uses the same address that was used to contact it in the login redirect. We could force it to use the address passed on start-up or add that as an option. But as CRI-O rewrote the address itself, it does have all the information it needs to figure out it's not a different address.
